agent/cache/watch_test.go

```diff
@@ -0,0 +1,165 @@
+		Meta: ResultMeta{Hit: false, Index: 5},
+		Err:  nil,
+	})
+
+	// We could wait for a full timeout but we can't directly observe it so
+	// simulate the behaviour by triggering a response with the same value and
+	// index as the last one.
+	close(trigger[1])
+
+	// We should NOT be notified about that. Note this is timing dependent but
+	// it's only a sanity check, if we somehow _do_ get the change delivered later
+	// than 10ms the next value assertion will fail anyway.
+	time.Sleep(10 * time.Millisecond)
+	require.Len(ch, 0)
+
+	// Trigger final update
+	close(trigger[2])
+
+	TestCacheWatchChResult(t, ch, WatchUpdate{
+		Result: 42,
+		Meta:   ResultMeta{Hit: false, Index: 7},
+		Err:    nil,
+	})
+
+	t.Log("Cancelling")
+	cancel()
+	t.Log("Cancelled")
+
+	// It's likely but not certain that the Watcher was already blocking on the
+	// next call. Since it won't timeout for 10 minutes, we can verify the
+	// cancellation worked by early terminating the call.
+	close(trigger[3])
+
+	time.Sleep(1 * time.Second)
+
+	// ch should now be closed (return zero value)
+	TestCacheWatchChResult(t, ch, WatchUpdate{})
+}
+
+// Test that a refresh performs a backoff.
+func TestCacheWatch_ErrorBackoff(t *testing.T) {
+	t.Parallel()
+
+	typ := TestType(t)
+	defer typ.AssertExpectations(t)
+	c := TestCache(t)
+	c.RegisterType("t", typ, &RegisterOptions{
+		Refresh: false,
+	})
+
+	// Configure the type
+	var retries uint32
+	fetchErr := fmt.Errorf("test fetch error")
+	typ.Static(FetchResult{Value: 1, Index: 4}, nil).Once()
+	typ.Static(FetchResult{Value: nil, Index: 5}, fetchErr).Run(func(args mock.Arguments) {
+		atomic.AddUint32(&retries, 1)
+	})
+
+	require := require.New(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	ch, err := c.Watch(ctx, "t", TestRequest(t, RequestInfo{Key: "hello"}))
+	require.NoError(err)
+
+	// Should receive the first result pretty soon
+	TestCacheWatchChResult(t, ch, WatchUpdate{
+		Result: 1,
+		Meta:   ResultMeta{Hit: false, Index: 4},
+		Err:    nil,
+	})
+
+	numErrors := 0
+	// Loop for a little while and count how many errors we see reported. If this
+	// was running as fast as it could go we'd expect this to be huge. We have to
+	// be a little careful here because the watch chan ch doesn't have a large
+	// buffer so we could be artificially slowing down the loop without the
+	// backoff actualy taking affect. We can validate that by ensuring this test
+	// fails without the backoff code reliably.
+	timeoutC := time.After(500 * time.Millisecond)
+OUT:
+	for {
+		select {
+		case <-timeoutC:
+			break OUT
+		case u := <-ch:
+			numErrors++
+			require.Error(u.Err)
+		}
+	}
+	// Must be fewer than 10 failures in that time
+	require.True(numErrors < 10, fmt.Sprintf("numErrors: %d", numErrors))
+
+	// Check the number of RPCs as a sanity check too
+	actual := atomic.LoadUint32(&retries)
+	require.True(actual < 10, fmt.Sprintf("actual: %d", actual))
+}
```

Comment:
Note: I did have another test here to test badly-behaved RPCs that return zero values, but it turns out that it only exercised the existing code in `Get` and failed for a subtle reason - the test was both returning 0 index AND never blocking even if the request came in with `index = 1`. This is not behaviour we expect of any RPC that uses blockingQuery helper _even_ if it's badly behaved and returns zero index. Fixing `Get` to handle taht would be a ton of refactoring for a case that can never realistically exist.

Stuff like this is a good example of why so many layers of blocking is hard to work with!